# Блок A2.5 Модуля A2

## А2.5.1 Использование словарей и списков для хранения данных
В предыдущей части мы изучили распространенные структуры данных такие, как списки, словари, а также различные производные конструкции, например, списки словарей, вложенные словари.

Рассмотренные нами структуры данных часто используются для хранения информации о каких-то событиях, происходящих внутри систем. Записи о событиях, которые совершаются в ходе выполнения программы или использования программы пользователями, чаще всего записываются в файлы, которые называются логами. В качестве примера мы рассмотрим небольшой кусочек логов мобильного приложения.

Логи хранятся в базах данных или в файлах. Мы рассмотрим случай, когда логи хранят в файлах.

## А2.5.1 Использование словарей и списков для хранения данных
В предыдущей части мы изучили распространенные структуры данных такие, как списки, словари, а также различные производные конструкции, например, списки словарей, вложенные словари.

Рассмотренные нами структуры данных часто используются для хранения информации о каких-то событиях, происходящих внутри систем. Записи о событиях, которые совершаются в ходе выполнения программы или использования программы пользователями, чаще всего записываются в файлы, которые называются логами. В качестве примера мы рассмотрим небольшой кусочек логов мобильного приложения.

Логи хранятся в базах данных или в файлах. Мы рассмотрим случай, когда логи хранят в файлах.

## А2.5.2 Первая работа с файлами
### Чтение логов из JSON-файла
Одним из форматов, в которых хранятся логи, является формат JSON. Этот формат позволяет хранить данные в структурном виде, то есть в виде списков, словарей, а также более сложных вложенных конструкций.
Скачать файл можно по ссылке: https://lms.skillfactory.ru/assets/courseware/v1/6d23dbcdd02e18bb40dcdee6cd4a94d5/asset-v1:Skillfactory+PY-30+16OCT2019+type@asset+block/data.json

Словарь содержит различные параметры, такие как:

* идентификатор события (id);
* идентификатор клиента (client_id);
* идентификатор пользователя (user_id);
* категория действия (category);
* действие, которое совершил пользователь (action);
* дополнительные параметры события (options).

Мы будем использовать не все параметры, а только client_id и category, но вы можете заняться более детальным анализом и исследовать эти данные, используя все параметры.

## А2.5.3 Открытие файла в Python
### Библиотека json и функция open
Для того чтобы открыть JSON-файл с помощью Python и прочитать данные, можно воспользоваться библиотекой JSON. Сейчас нам не стоит заострять внимание на особенностях её использования. Главное —  научиться открывать с её помощью файлы. Делается это через следующую конструкцию:

In [2]:
import json
with open('data.json', 'rb') as infile:
    data = json.load(infile)

Здесь 'data.json' указывает на путь к файлу. Чтобы код сработал корректно, нужно чтобы файл с данными хранился в той же папке, в которой вы создали ноутбук с кодом. 

Итак, этот код открывает файл 'data.json' и затем преобразует JSON в объекты, с которыми может работать Python: в словари и списки. Всё это записывается в переменную data.

Давайте посмотрим, что представляет собой data:

In [3]:
type(data)

dict

Эта переменная представляет собой словарь. Значит, у него есть ключи. Посмотрим их:

In [4]:
data.keys()

dict_keys(['events_data'])

В нашем словаре всего один ключ 'events_data'. Посмотрим, какое значение доступно по этому ключу:

In [5]:
data['events_data']

[{'id': 47946124,
  'client_id': 62526,
  'user_id': 110777,
  'category': 'page',
  'action': 'enter',
  'options': {'name': 'landing'}},
 {'id': 47947076,
  'client_id': 61944,
  'user_id': 108565,
  'category': 'page',
  'action': 'enter',
  'options': {'name': 'landing'}},
 {'id': 47962056,
  'client_id': 27115,
  'user_id': 49440,
  'category': 'page',
  'action': 'enter',
  'options': {'name': 'page'}},
 {'id': 47962120,
  'client_id': 27115,
  'user_id': 49440,
  'category': 'datepicker',
  'action': 'apply',
  'options': {'endDate': '2019-03-31 23:59:59',
   'startDate': '2019-03-25 00:00:00'}},
 {'id': 47962218,
  'client_id': 27115,
  'user_id': 49440,
  'category': 'table',
  'action': 'sort',
  'options': {'dir': 'ASC', 'name': 'deal_date'}},
 {'id': 47962238,
  'client_id': 27115,
  'user_id': 49440,
  'category': 'table',
  'action': 'sort',
  'options': {'dir': 'DESC', 'name': 'deal_date'}},
 {'id': 47962532,
  'client_id': 27115,
  'user_id': 49440,
  'category': 'table

Мы получили список, в котором хранятся наши события в виде отдельных словарей. С этими словарями мы уже познакомились выше, когда рассматривали структуру нашего JSON-файла с логом.

## А2.5.4 Работа с данными
### Покопаемся в данных
Для удобства давайте запишем полученный в предыдущем модуле список в новую переменную data_list, чтобы нам не приходилось каждый раз обращаться к одному и тому же ключу 'events_data' в словаре data.

In [6]:
data_list = data['events_data']

Теперь мы можем оценить, сколько событий хранится в нашем логе. Для этого посчитаем длину получившегося списка:

In [7]:
print(len(data_list))

107


Целых 107 событий! В первую очередь, нам интересно знать, какие это события. Для этого давайте пройдемся циклом по всем событиям и запишем в новый список все встречающиеся категории (ключ 'category'):

In [8]:
categories = []
for item in data_list:
    category = item['category']
    categories.append(category)
print (categories)

['page', 'page', 'page', 'datepicker', 'table', 'table', 'table', 'table', 'datepicker', 'datepicker', 'datepicker', 'datepicker', 'datepicker', 'page', 'page', 'page', 'page', 'page', 'page', 'page', 'page', 'page', 'page', 'page', 'page', 'page', 'page', 'page', 'page', 'page', 'page', 'page', 'page', 'report', 'report', 'report', 'report', 'page', 'report', 'page', 'page', 'page', 'table', 'table', 'table', 'table', 'page', 'page', 'page', 'page', 'page', 'page', 'page', 'page', 'page', 'page', 'page', 'page', 'report', 'page', 'page', 'page', 'page', 'page', 'report', 'page', 'page', 'page', 'page', 'page', 'page', 'page', 'page', 'page', 'page', 'page', 'page', 'page', 'datepicker', 'datepicker', 'datepicker', 'datepicker', 'datepicker', 'datepicker', 'datepicker', 'page', 'page', 'page', 'page', 'page', 'page', 'report', 'table', 'table', 'datepicker', 'datepicker', 'page', 'page', 'page', 'page', 'page', 'page', 'page', 'page', 'page', 'page', 'page']


Как видно, чаще всего встречается категория 'page', означающая просмотр страницы.

Давайте подсчитаем частоту встречаемости каждого из событий. Для этого можно использовать удобную конструкцию Counter, которая доступна в библиотеке Collections. Эта библиотека идёт в составе Python, то есть её не нужно устанавливать дополнительно, она предоставляет широкие возможности для работы с различными структурами данных. 

In [9]:
import collections
c = collections.Counter()
for category in categories:
    c[category] += 1
print (c)

Counter({'page': 74, 'datepicker': 15, 'table': 10, 'report': 8})


Таким образом мы посчитали сколько разных событий записано в наш лог.

## А2.5.5 Анализируем данные
### Первый пример анализа данных
Теперь давайте посмотрим, какие клиенты совершают события table. Для этого пройдемся по каждому словарю из списка data_list и добавим значение client_id в новый список table_clients, но только в тех случаях, где category = 'table':

In [10]:
table_clients = []
for item in data_list:
    client_id = item['client_id']
    category = item['category']
    if category == 'table':
        table_clients.append(client_id)
print (table_clients)

[27115, 27115, 27115, 27115, 60459, 60459, 60459, 60459, 58113, 58113]


У нас получился список table_clients, который включает в себя клиентов, совершавших действие с категорией 'table'. Давайте применим такую же конструкцию, которую мы применяли для подсчета количества разных событий, чтобы посчитать, кто из клиентов чаще всего совершал действие 'table':

In [11]:
c = collections.Counter()
for table_client in table_clients:
    c[table_client] += 1
print (c)

Counter({27115: 4, 60459: 4, 58113: 2})


Мы видим, что у нас всего 3 пользователя, два из которых совершили по 4 действия.

У нас данных мало, поэтому всё можно оценить просто выводом словаря «c» на экран с помощью print, но иногда данных так много, что посчитать количество пользователей не так просто. Для этого можно посчитать, сколько ключей содержится в словаре «c»:

In [12]:
print (len(c.keys()))

3


Вывести информацию по конкретному пользователю можно, обратившись по соответствующему ключу:

In [13]:
print(c[27115])

4


Итак, мы рассмотрели пример анализа данных с помощью использования словарей и списков. Иногда это действительно удобный способ, но в дальнейшем мы рассмотрим применение пакета Pandas для анализа данных, который раскроет нам новые возможности. С помощью этого пакета также можно анализировать данные, но при этом в более гибкой и удобной форме.

## А2.5.6 Задания

1. Подсчитайте количество клиентов (client_id), которые совершали какие-либо действия

In [14]:
c = collections.Counter()
for item in data_list:
    c[item['client_id']]+=1
print(len(c))

51


* 49
* 51 [верно]
* 50

2. Сколько действий совершил клиент (client_id) под номером 60459?

In [19]:
c = collections.Counter()
kolvo=0
for item in data_list:
    if item['client_id']==60459:
        kolvo+=1 # Увеличиваем счетчик количества действий определенного клиента
print(kolvo)

12


3. Сколько действий с категорией (category) = page совершил клиент под номером 62602?

In [23]:
kolvo=0
for item in data_list:
    if item['client_id']==62602:
        if item['category']=='page':
            kolvo+=1
print(kolvo)

9


4. Сколько уникальных клиентов совершили действия с категорией (category) = report?


In [25]:
c=collections.Counter()
new=[]
for item in data_list:
    category=item['category']
    if category=='report':
        new.append(item['client_id']) #Добавляем в новый словрь всех клиентов с данной категорией
for s in new:
    c[s]+=1 # считаем уникальное число клиентов в словаре
print(len(c))

3


5. Перечислите в каждом из полей идентификаторы клиентов, которые совершили действия с категорией (category) = report (В порядке возрастания)

In [26]:
c=collections.Counter()
new=[]
for item in data_list:
    if item['category']=='report':
        new.append(item['client_id'])
for s in new:
    c[s]+=1
print(c)

Counter({60459: 5, 62602: 2, 58113: 1})
